### ISSUE TITLE
[REVIEW]: qgs: A flexible Python framework of reduced-order multiscale climate models

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2597]
**Submitting author:** @jodemaey (<a href="http://orcid.org/0000-0002-5098-404X">Jonathan Demaeyer</a>)
**Repository:** <a href="https://github.com/Climdyn/qgs" target ="_blank">https://github.com/Climdyn/qgs</a>
**Version:** v0.2.0
**Editor:** @harpolea
**Reviewer:** @eviatarbach, @sadielbartholomew
**Archive:** <a href="https://doi.org/10.5281/zenodo.4368844" target="_blank">10.5281/zenodo.4368844</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/3cf640133997e1394e3fd22963730391/status.svg)](https://joss.theoj.org/papers/3cf640133997e1394e3fd22963730391)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/3cf640133997e1394e3fd22963730391"><img src="https://joss.theoj.org/papers/3cf640133997e1394e3fd22963730391/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/3cf640133997e1394e3fd22963730391/status.svg)](https://joss.theoj.org/papers/3cf640133997e1394e3fd22963730391)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/Climdyn/qgs" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@eviatarbach & @sadielbartholomew, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @harpolea know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/Climdyn/qgs">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@jodemaey) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/Climdyn/qgs">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@jodemaey) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
PDF failed to compile for issue #2597 with the following error: 

 Can't find any papers to compile :-(

**Comment by harpolea** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02597/joss.02597/10.21105.joss.02597.pdf)

**Comment by eviatarbach** [comment]
qgs provides a modular implementation of a quasi-geostrophic atmospheric model. Having previously used the MAOOAM model in research, I was impressed with its ease-of-use, highly readable code, the elegance of the tensor-based implementation, and built-in tangent linear and adjoint models. qgs builds on this by adding two different land components, and making it easy to switch between ocean and land coupling. qgs is also written entirely in Python, as opposed to MAOOAM that had Fortran and Lua implementations also. I commend the authors for their excellent contribution and hope that qgs will become a widely used tool in the modelling hierarchy.

Suggested changes:
1. There are a few missing parts of the documentation, as per the review checklist. The statement of need is provided in the software paper, but it has to be in the documentation as well.
> - A statement of need: Do the authors clearly state what problems the software is designed to solve and who the target audience is?
> - Community guidelines: Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support
2. I would like to see the software paper mention the availability of the tangent linear and adjoint models, as this is a major advantage for data assimilation research, as well as for, e.g., computing the Lyapunov spectrum. Adding an example of the usage of the tangent linear or adjoint models to the documentation would be helpful.
3. This implementation is written in Python instead of Fortran, but uses Numba and sparse matrices to accelerate the computations. I would thus like to see a performance comparison to the Fortran version of MAOOAM, either in the documentation or the software paper (see the performance item on the checklist).

Minor issues:
- "Simpy" should be Sympy in the forthcoming developments section.
- The link to the notebooks folder works in the GitHub README, but is broken in the readthedocs documentation. Also, the link to the main folder just goes to the main page of the documentation.
- "montain" -> "mountain" in the Reinhold and Pierrehumbert (1982) example page

**Comment by kthyng** [comment]
Hi @jodemaey! I am the rotating associate editor in chief this week and am going through stale submissions. I see that one of your reviewers put forth some comments above but you haven't responded yet. Will you be able to get to this soon? If not, that is ok, but we should in that case pause this submission. I recommend that we pause this if we haven't heard back from the author within a week or so.

**Comment by sadielbartholomew** [comment]
Regarding your comment @kthyng, just to clarify that I was waiting on the submitting author's response to the first reviewers' comments, as in https://github.com/openjournals/joss-reviews/issues/2597#issuecomment-683972381, before starting my review, in case there were any significant changes made to the codebase or paper based on that feedback.

However, as you state it has been a while now, a good month, since that feedback without response. Would you recommend I start my review regardless, now? Or wait considering this could be considered 'stale'? I am just wary we are approaching six weeks since the review was opened and it is recommended I complete my review within that timeframe.

**Comment by jodemaey** [comment]
Hi @kthyng and @sadielbartholomew, 

There was a misunderstanding here. I was waiting for the second review to address all the comments at once.

In any case I'm not able to work on this until the end of October. At that time I will work again fully on qgs and will address the comments and suggestions of @eviatarbach .

**Comment by sadielbartholomew** [comment]
Sorry that there was this misunderstanding, @jodemaey.

If you are unable to make changes until November it sounds like it would be best for me to complete my review by then so that you can know what else, if anything, might need to be addressed. I will ensure I have my review completed by then, and try to do it much sooner anyhow, now that I know about the situation. Thanks.

**Comment by kthyng** [comment]
@jodemaey, @sadielbartholomew I'm glad I commented then to clear this up! 

@sadielbartholomew sounds good to me. I will not pause this submission then since we'll expect to hear from you in the next month with your review, and then subsequently from @jodemaey with responses. 

Thanks all!

**Comment by jodemaey** [comment]
@sadielbartholomew , I am also sorry about this, I should have replied to Eviatar's review to clarify this. Thank you in advance for your review.

@kthyng Thank you.

**Comment by arfon** [comment]
:wave: @sadielbartholomew - just checking you're planning on completing your review in the next week or so?

**Comment by sadielbartholomew** [comment]
Hi @arfon, yes, I will complete a first review over the weekend. Thanks for your patience.

**Comment by sadielbartholomew** [comment]
Really sorry for the further delay, I am having to co-manage an urgent operational infrastructure situation. I should have some time to review this later tonight and tomorrow so can complete my review by Saturday morning (UK time).

**Comment by sadielbartholomew** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List Whedon's capabilities
@whedon commands

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

EDITORIAL TASKS

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Ask Whedon to check the references for missing DOIs
@whedon check references

# Ask Whedon to check repository statistics for the submitted software
@whedon check repository

```

**Comment by sadielbartholomew** [comment]
@whedon check repository

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.84  T=0.36 s (159.6 files/s, 37541.5 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
SVG                              4              4              4           2937
Python                          17           1058           1792           2480
reStructuredText                20            373            387            454
Jupyter Notebook                 8              0           3037            439
TeX                              1             34              0            385
Markdown                         1             46              0             97
DOS Batch                        1              8              1             26
YAML                             2              5              4             23
Bourne Shell                     3              8              0             15
make                             1              4              9              9
-------------------------------------------------------------------------------
SUM:                            58           1540           5234           6865
-------------------------------------------------------------------------------


Statistical information for the repository 'ce242648c96bcd1558042443' was
gathered on 2020/11/06.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Jonathan Demaeyer               18          6983           1653          100.00

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Jonathan Demaeyer          5330           76.3          4.0                6.45
```

**Comment by jodemaey** [comment]
Hi @sadielbartholomew ,

I have seen that you raised two issues in the qgs repo. Does that mean that your review is complete?

Thank you in advance,

Jonathan

**Comment by sadielbartholomew** [comment]
Hi Jonathan, no I only got part way through my review in the time I could devote last week. I will try to finish the review tonight and tomorrow. Sorry for the delay and thanks for your patience.

**Comment by arfon** [comment]
:wave: @sadielbartholomew - just checking in here to see how you're getting on?

**Comment by sadielbartholomew** [comment]
Sorry @arfon I still need a bit more time here. Let me promise that by the end of this week (Sunday 29th, UK time) I will finish the review.

**Comment by arfon** [comment]
> Sorry @arfon I still need a bit more time here. Let me promise that by the end of this week (Sunday 29th, UK time) I will finish the review.

:+1: no problem. Thanks for the update!

**Comment by jodemaey** [comment]
> Sorry @arfon I still need a bit more time here. Let me promise that by the end of this week (Sunday 29th, UK time) I will finish the review.

Hi @sadielbartholomew ,

Thank you. Meanwhile, I've addressed your two issues on the `qgs` repo.

**Comment by sadielbartholomew** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02597/joss.02597/10.21105.joss.02597.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02597/joss.02597/10.21105.joss.02597.pdf) :page_facing_up: :point_left:

**Comment by sadielbartholomew** [comment]
Hi @jodemaey, I assume that the current state of the proposed JOSS paper content is contained in the `joss` branch rather than `master`? I've just re-generated it via the whedon tool, as per the above comments.

Please in future let us know when you update the paper, since I just spent a significant amount of time reviewing the old version only to see that it has changed quite a bit now, with a new section (that's okay, but going forward it helps for us to know of any state changes).

It's good in the respect that you have, with the updates since the last version generated and linked in this thread (https://github.com/openjournals/joss-reviews/issues/2597#issuecomment-679154289), addressed some aspects that I was going to raise, e.g. interlinking between the references in the text and their corresponding entries in the 'References' section.

However there are some minor aspects of the updated paper itself that I have feedback on. Since these are spread across the `paper.md` , it is clearest if I can refer to specific lines there, so I will use a Pull Request (which I'll open in a moment and link to this Issue) to provide this feedback. Feel free to close it or merge it, as you wish, but consider the contained changeset and comments upon it formal feedback on the paper content (i.e. towards the items listed uner 'Software paper' in the review checklist).

Thanks.

**Comment by jodemaey** [comment]
Hi @sadielbartholomew ,

Thanks for your suggestions. I will merge them to the joss branch when your review is finished.

**Comment by jodemaey** [comment]
Hi @sadielbartholomew ,

I had a look at your commit. Can I merge it as it is?  What about the other items on the checklist about functionalities and documentation: have you considered them already?

**Comment by sadielbartholomew** [comment]
> I had a look at your commit. Can I merge it as it is?

Hi @jodemaey, sorry I fell asleep before I put it in as a PR. Please don't merge it yet. Let me open it now since many of my comments are not contained in the commit but I intended to put as comments on said PR. One moment.

> What about the other items on the checklist about functionalities and documentation: have you considered them already?

I am nearly ready to tick these but not quite since I want to be thorough. I'll let you know by the end of the day. Thanks.

**Comment by sadielbartholomew** [comment]
Just to note here that I am ticking off items from the review checklist based on the `joss` branch, rather than the `master` branch, as that has been updated in response to feedback and will be the branch used for the state of the codebase for the archive and for the paper content.

Notably, for reviewing the items under 'Documentation' I have built the documentation locally from the `joss` branch (via the documented procedure); I am not referring to the documentation currently hosted online.

**Comment by sadielbartholomew** [comment]
@jodemaey: regarding the installation, the documented approach to installation does work in that it allows the scripts to run successfully, so that is sufficient for the review checklist item. Saying that, I've raised an issue because going forward ideally the package could be made available to install from `PyPI` or some `conda` channel or similar.

I do however have to query a claim in the documentation on that note, namely as stated under the `general_information.html#installation` endpoint of the documentation (from the `joss` branch so built locally):

> The easiest way to install is through Anaconda.

The installation documented there isn't by Anaconda, as conda is only being used to create and activate an environment with the required dependencies. Strictly nothing is being 'installed' at all as no package has been set up, rather we can run `qgs` as documented because the `python` interpreter is being run directly on the set of top-level scripts.

I think the wording of the documentation in this respect needs to be amended to reflect the fact that there is no 'installation' going on per se, else parts such as that sentence are confusing and untrue. An environment is set up with the dependencies, then the scripts, cloned locally, can be run. So for example, the quoted sentence above would in accuracy be something like "The easiest way to run qgs is to use an appropriate environment created through Anaconda" or similar.

**Comment by jodemaey** [comment]
@sadielbartholomew : I have responded to your remark about packaging the code on the `qgs` repo issues page: https://github.com/Climdyn/qgs/issues/4

**Comment by jodemaey** [comment]
Hi @sadielbartholomew , 

Please note that new modifications of the paper have been pushed to the joss branch in commit: https://github.com/Climdyn/qgs/commit/b19cbadc8020f8f77f7e451f35d922ff6d1a935c

**Comment by jodemaey** [comment]
Hi @sadielbartholomew ,

When can I reasonably expect your review to be final? I need that information to plan my week of work.

Thank you in advance.

**Comment by sadielbartholomew** [comment]
Hi @jodemaey, I'm not far off with finalising the review so will try tonight (early hours of Tuesday UTC) to report on anything I consider outstanding towards acceptance. If not, I'll definitely be done by Wednesday night. Thanks again for your patience.

**Comment by sadielbartholomew** [comment]
Thanks for opening up an issue indicating you are going to address my previous comment regarding the installation instructions, @jodemaey. I was going to re-raise that in case you missed it. Otherwise all of my feedback raised so far has been addressed well, bar one small formatting change to the text, described against point (3) below.

I have now completed the remainder of my review and I've outlined the results below (in short I'm happy with everything except for a small number of items I feel still need to be addressed). Thanks for your patience waiting for the rest of the review.


### Final review comments

If the following three points are addressed satisfactorily then I am more than happy to tick the remaining items in the review checklist and moreover to recommend acceptance of the `qgs` paper to JOSS:

1. **Regarding the state of the field**, the review criterion is to 'describe how this software compares to other commonly-used packages' and whilst several similar packages are currently referenced and briefly described under a section of that name, I can't find any *comparison* between qgs and any of those packages in the text. There is some indication of the relation to MAOOAM, but since that is your own package and is included in qgs, I wouldn't count it as a separate package.

    So, what I would like to see towards fulfilling that criterion is some brief comparison between qgs and the listed alternative "easy-to-use idealized atmospheric model[s]" as you introduce them (i.e. q-gcm, pyqg, Isca). A sentence or two on what qgs offers that these libraries do not, or vice versa, would suffice. It may tie in with ground you have already covered in the first two paragraphs of the 'Statement of need' section, so you could perhaps reference them there with some further detail to address this point.

2. **The functionality** of the software is documented really well, which is great. I did feel that the API reference documentation is quite hard to find, though. For example, to see the model configuration classes I would have to go through the menu like so: 'Code Description -> Modules -> Configuration of the model' and then there's a bit of introductory text before the class signature and parameters are listed, so if a reader was viewing on a laptop they might not see it and not realise the classes were covered below.

    In my experience of using and creating software documentation, it is most common to have the API reference as a top-level menu item, under a name such as 'API' or 'reference', where it is evident and easy to access. Therefore I advise you:
    * move the 'Modules' section out to a higher-level in the documentation tree; and/or (but ideally 'and')
    * rename it to one of the names above, as 'Modules' could imply an overview or commentary on the modules rather than the API reference for them.

3. **For the paper text formatting**, please convert `python` to `Python` in the Performance section of the paper text (as raised in Climdyn/qgs#3 but not yet addressed).

Finally, I have **some further comments that are not compulsory for my review acceptance**, but that I think would improve the software or the paper content:

4. I notice the labels on both the bar charts, (a) and (b), in Figure 1  are identical, so to make the visual easier to comprehend I suggest removing the labels on the top chart and moving it so the bottom is joined to the top of the lower chart, or to convert it into a single bar chart with two bars per item on the x-axis.

5. I've noticed that the repository is structured in a way that makes it difficult to search through, namely with the `qgs` functional modules all in separate directories at the root of the repo, instead of being contained within a directory underneath the root which is much more common. This means one can't use a simple `git grep` to search for keywords in the modules to aid with development, as it will pick up words from the Notebook `.ipynb` files (which are big JSON structures with lots of binary code to save the images etc. so clog up the terminal on a search) or the documentation source and build pages, etc., too.

    For that reason, so any contributors can easily work on the core code, personally I'd move all of the module directories under another dedicated directory called `qgs` or `lib` or similar.

**Comment by jodemaey** [comment]
Dear all,

Please find below our response to the reviews of the referees (see the posts https://github.com/openjournals/joss-reviews/issues/2597#issuecomment-683972381 and https://github.com/openjournals/joss-reviews/issues/2597#issuecomment-741610669 ).

# Response to referee 1

We thank @eviatarbach for his helpful review. We respond below to the suggested changes:

> qgs provides a modular implementation of a quasi-geostrophic atmospheric model.
> Having previously used the MAOOAM model in research, I was impressed with its
> ease-of-use, highly readable code, the elegance of the tensor-based
> implementation, and built-in tangent linear and adjoint models. qgs builds on
> this by adding two different land components, and making it easy to switch
> between ocean and land coupling. qgs is also written entirely in Python, as
> opposed to MAOOAM that had Fortran and Lua implementations also. I commend the
> authors for their excellent contribution and hope that qgs will become a widely
> used tool in the modelling hierarchy.
>
> Suggested changes:
>
>    1. There are a few missing parts of the documentation, as per the review
>    checklist. The statement of need is provided in the software paper, but it has
>     to be in the documentation as well.
>      * A statement of need: Do the authors clearly state what problems the software is
>       designed to solve and who the target audience is?    
>      * Community guidelines:
>      Are there clear guidelines for third parties wishing to 1) Contribute to the
>      software 2) Report issues or problems with the software 3) Seek support

We have added a *statement of need* section in the documentation. We have also added community guidelines on how to 
contribute to the development. These additions are contained in the commit f87022c6c1f5d5909129fdbba029566f07b1d2bb in the `joss` branch.

>    2. I would like to see the software paper mention the availability of the
>    tangent linear and adjoint models, as this is a major advantage for data
>    assimilation research, as well as for, e.g., computing the Lyapunov spectrum.
>    Adding an example of the usage of the tangent linear or adjoint models to the
>     documentation would be helpful.

We now mention the tangent linear explicitly in the paper. We have also added a usage example in the documentation. 
Thank you for the suggestion. The commits of these additions are cee79b7b53dcfe7d80b535d2badb0b678aeb4761 and 9fd689d8ab2bfcdd548f708e8862c76ae6c706cb .
>    3. This implementation is written in Python instead of Fortran, but uses Numba
>    and sparse matrices to accelerate the computations. I would thus like to see a
>    performance comparison to the Fortran version of MAOOAM, either in the
>    documentation or the software paper (see the performance item on the
>     checklist).

We have now made a benchmark of `qgs` versus MAOOAM lua/fortran code. The results are detailed in a new section of the paper.
   Related commits: dfaeed977de69ef518474db6f2458e6312bf9938 .
> Minor issues:
>    * "Simpy" should be Sympy in the forthcoming developments section.
>    * The link to the notebooks folder works in the GitHub README, but is broken in
>    the readthedocs documentation. Also, the link to the main folder just goes to
>     the main page of the documentation.
>    * "montain" -> "mountain" in the Reinhold and Pierrehumbert (1982) example page
We have also addressed the raised minor issues in the commit bac1d3c67a383b041b686a6c8bc903d159d602b6 .

We thank Eviatar Bach again and hope that the proposed modifications address all the concerns. 

# Response to referee 2

We thank @sadielbartholomew for her thorough review.
We address here her final comments, the other ones have already been addressed in the discussion above.

> Thanks for opening up an issue indicating you are going to address my previous
> comment regarding the installation instructions, @jodemaey.
> I was going to re-raise that in case you missed it. Otherwise
> all of my feedback raised so far has been addressed well, bar one small
> formatting change to the text, described against point (3) below.
>
> I have now completed the remainder of my review and I've outlined the results
> below (in short I'm happy with everything except for a small number of items I
> feel still need to be addressed). Thanks for your patience waiting for the rest
> of the review. Final review comments
>
> If the following three points are addressed satisfactorily then I am more than
> happy to tick the remaining items in the review checklist and moreover to
> recommend acceptance of the qgs paper to JOSS:
>
> 1. Regarding the state of the field , the review criterion is to 'describe how this
> software compares to other commonly-used packages' and whilst several similar
> packages are currently referenced and briefly described under a section of that
> name, I can't find any comparison between qgs and any of those packages in the
> text. There is some indication of the relation to MAOOAM, but since that is
> your own package and is included in qgs, I wouldn't count it as a separate
> package. So, what I would like to see towards fulfilling that criterion is some brief
> comparison between qgs and the listed alternative "easy-to-use idealized
> atmospheric model[s]" as you introduce them (i.e. q-gcm, pyqg, Isca). A
> sentence or two on what qgs offers that these libraries do not, or vice versa,
> would suffice. It may tie in with ground you have already covered in the first
> two paragraphs of the 'Statement of need' section, so you could perhaps
> reference them there with some further detail to address this point.

 We have better described the other models in this list and have stressed the main assets of `qgs` in this regard.
   Related commits: 94188990ecbddda0409a98f820c34467962b0eab , 0ff4d83621ffa057722856d39e5141895f7ec32c and 5723bebe6d00dd0354ce70f1f26ae1b81a54af6b .

> 2. The functionality of the software is documented really well, which is great. I
> did feel that the API reference documentation is quite hard to find, though.
> For example, to see the model configuration classes I would have to go through
> the menu like so: 'Code Description -> Modules -> Configuration of the model'
> and then there's a bit of introductory text before the class signature and
> parameters are listed, so if a reader was viewing on a laptop they might not
> see it and not realise the classes were covered below.
> In my experience of using and creating software documentation, it is most common
> to have the API reference as a top-level menu item, under a name such as 'API'
> or 'reference', where it is evident and easy to access. Therefore I advise you:
>   * move the 'Modules' section out to a higher-level in the documentation tree;
>     and/or (but ideally 'and')
>   * rename it to one of the names above, as 'Modules' could imply an overview or
>     commentary on the modules rather than the API reference for them.

We agree that the proposed change will increase the readability of the documentation. Thank you. 
We have implemented it in the commit 7bc75cf9f1609d850ea6cc04939c8ce24b8ff912.

> 3. For the paper text formatting , please convert "python" to "Python" in the
> Performance section of the paper text (as raised in [Climdyn/qgs#3](https://github.com/Climdyn/qgs/pull/3) but not yet addressed).

This was done in the commit 4c1e0a85f4997aeb614cd926ed51f5cb06b9306f .
  
> Finally, I have some further comments that are not compulsory for my review
> acceptance , but that I think would improve the software or the paper content:
> 1. I notice the labels on both the bar charts, (a) and (b), in Figure 1 are
> identical, so to make the visual easier to comprehend I suggest removing the
> labels on the top chart and moving it so the bottom is joined to the top of the
> lower chart, or to convert it into a single bar chart with two bars per item on
> the x-axis.

We do not think that this modification would make the figure easier to read. Creating a single plot could indeed be a good solution,
   but the data of the two actual panels are too different and that would decrease the readability of the panel *(a)* data.
  
> 2. I've noticed that the repository is structured in a way that makes it difficult
> to search through, namely with the qgs functional modules all in separate
> directories at the root of the repo, instead of being contained within a
> directory underneath the root which is much more common. This means one can't
> use a simple git grep to search for keywords in the modules to aid with
> development, as it will pick up words from the Notebook .ipynb files (which are
> big JSON structures with lots of binary code to save the images etc. so clog up
> the terminal on a search) or the documentation source and build pages, etc.,
> too.
> For that reason, so any contributors can easily work on the core code,
> personally I'd move all of the module directories under another dedicated
> directory called qgs or lib or similar.
>


Thank you for this suggestion. We could consider changing the structure of the repository to indeed simplify it along the line that you propose.
However, we do not want to make that change now, since some consequent refactorings are planned for the v0.3. We need to think about it, and as such
we have opened the issue https://github.com/Climdyn/qgs/issues/7 to discuss and come back to it later.


We thank Sadie Bartholomew again for her suggestions.

# General comment

Various small language changes and consistency have also been applied.
Please see the commits f8d8e41c2a6e2b676ee242a8f55e791cc1de303e and 
a318efbafdd4324ffb567ef70171ab222aae80d9 for a recap.

**Comment by jodemaey** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02597/joss.02597/10.21105.joss.02597.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02597/joss.02597/10.21105.joss.02597.pdf) :page_facing_up: :point_left:

**Comment by arfon** [comment]
Thanks for the detailed response @jodemaey.

@eviatarbach, @sadielbartholomew - when you get a moment, please come and take a look at the author responses and let us know what you think (updating the last few checkboxes too if appropriate).

**Comment by sadielbartholomew** [comment]
Thanks @jodemaey for the considered response. **I have inspected your review response comment and the referenced commits and I am now satisfied that all of the review checklist items are covered well by the `qgs` code and the proposed paper and am delighted to recommend that the `qgs` paper be accepted.**

Just a few closing comments, only the first of which calls for any action (ideally it could be addressed before the paper is released should everyone else is happy for it to be accepted).

I rebuilt the documentation locally to check on the restructure for the API reference and there is now a top-level 'References' section which addresses the feedback at hand, but there seems to be a setup issue since many (but not all) of the pages, notably:

* Inner Products (living at `<path>/documentation/build/html/files/technical/inner_products.html`)
* Tensors module (`.../tensors.html`)
* Functions module (etc.)
* Integration modules

all seem to be blank, whereas in the documentation currently live on `qgs.readthedocs.io` those pages appear to be populated with the reference information. So that should be addressed to ensure the latest documentation retains all the reference information.

> We do not think that this modification would make the figure easier to read. Creating a single plot could indeed be a good solution,
but the data of the two actual panels are too different and that would decrease the readability of the panel (a) data.

That is fair enough. The plots are certainly readable as they are.

> Thank you for this suggestion. We could consider changing the structure of the repository to indeed simplify it along the line that you propose.
However, we do not want to make that change now, since some consequent refactorings are planned for the v0.3. We need to think about it, and as such
we have opened the issue Climdyn/qgs#7 to discuss and come back to it later.

I'm glad to hear you are considering this, of course address it at a time which is most appropriate in your planned roadmap.

**Comment by jodemaey** [comment]
Hi @sadielbartholomew ,

> I rebuilt the documentation locally to check on the restructure for the API reference and there is now a top-level 'References' section which addresses the feedback at hand, but there seems to be a setup issue since many (but not all) of the pages, notably:
> 
>     * Inner Products (living at `<path>/documentation/build/html/files/technical/inner_products.html`)
> 
>     * Tensors module (`.../tensors.html`)
> 
>     * Functions module (etc.)
> 
>     * Integration modules
> 
> 
> all seem to be blank, whereas in the documentation currently live on `qgs.readthedocs.io` those pages appear to be populated with the reference information. So that should be addressed to ensure the latest documentation retains all the reference information.
> 

I cannot reproduce your problem, on my system the documentation pages that you mention are fine.
Could you first perform a `make clean` before doing a `make html` again?

**Comment by sadielbartholomew** [comment]
Hi @jodemaey, thanks for confirming you can't recreate it. It was indeed my fault, sorry, I didn't realise but I built the local documentation last time with a different environment than the `qgs` conda one I used for reviewing, such that I did not have the right dependencies in my environment for Sphinx to use to parse the code to generate the API reference. I just tried it with the right environment activated and everything was there as it should be.

So everything is great now as far as I am concerned. Thanks again.

**Comment by eviatarbach** [comment]
The authors have addressed all my comments, and I am happy to recommend publication.

**Comment by arfon** [comment]
@whedon check references from branch joss

**Comment by whedon** [comment]
```
Attempting to check references... from custom branch joss
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1016/j.physd.2015.07.006 is OK
- 10.5194/gmd-9-2793-2016 is OK
- 10.1007/s13351-018-8012-y is OK
- 10.1145/2833157.2833162 is OK
- 10.5194/npg-27-307-2020 is OK
- 10.1175/1520-0469(1963)020<0130:DNF>2.0.CO;2 is OK
- 10.1017/CBO9780511617652.004 is OK
- 10.1111/j.1600-0870.1984.tb00230.x is OK
- 10.1038/s41586-020-2649-2 is OK
- 10.1038/s41592-019-0686-2 is OK
- 10.3233/978-1-61499-649-1-87 is OK
- 10.5334/jors.151 is OK
- 10.1029/2019MS001652 is OK
- 10.1007/s10955-020-02525-z is OK
- 10.1002/qj.3594 is OK
- 10.1007/s00382-020-05313-3 is OK
- 10.5194/gmd-11-843-2018 is OK
- 10.3402/tellusa.v41i4.11842 is OK
- 10.1504/IJCSE.2009.029165 is OK

MISSING DOIs

- 10.2142/biophys.48.324 may be a valid DOI for title: Isca
- 10.1017/cbo9780511617652.004 may be a valid DOI for title: Predictability: A problem partly solved

INVALID DOIs

- 10.1175/1520-0469(1987)044%3C3282:TOMEAW%3E2.0.CO;2 is INVALID
- 10.1175/1520-0493(1982)110%3C1105:DOWRQS%3E2.0.CO;2 is INVALID
- 10.1002/(SICI)1097-024X(199606)26:6%3C635::AID-SPE26%3E3.0.CO;2-P is INVALID
```

**Comment by arfon** [comment]
@jodemaey - please ignore these suggestions from Whedon, it looks like these DOIs are fine (although the URLs are somewhat unusual).

```
INVALID DOIs

- 10.1175/1520-0469(1987)044%3C3282:TOMEAW%3E2.0.CO;2 is INVALID
- 10.1175/1520-0493(1982)110%3C1105:DOWRQS%3E2.0.CO;2 is INVALID
- 10.1002/(SICI)1097-024X(199606)26:6%3C635::AID-SPE26%3E3.0.CO;2-P is INVALID
```

Please check though whether these DOIs are correct, and if so, please add them to your BibTeX:

```
MISSING DOIs

- 10.2142/biophys.48.324 may be a valid DOI for title: Isca
- 10.1017/cbo9780511617652.004 may be a valid DOI for title: Predictability: A problem partly solved
```

**Comment by arfon** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02597/joss.02597/10.21105.joss.02597.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02597/joss.02597/10.21105.joss.02597.pdf) :page_facing_up: :point_left:

**Comment by arfon** [comment]
Also, please give the PDF above a good read and check for any last fixes that might need to be made (I'll do the same).

**Comment by arfon** [comment]
@jodemaey - I also found this small fix for a formatting issue with one reference: https://github.com/Climdyn/qgs/pull/8

**Comment by arfon** [comment]
@jodemaey - after you've fixed up the DOIs and merged https://github.com/Climdyn/qgs/pull/8, could you make a new release of this software that includes the changes that have resulted from this review. Then, please make an archive of the software in Zenodo/figshare/other service and update this thread with the DOI of the archive? For the Zenodo/figshare archive, please make sure that:

- The title of the archive is the same as the JOSS paper title
- That the authors of the archive are the same as the JOSS paper authors

I can then move forward with accepting the submission.

**Comment by jodemaey** [comment]
> 
> Please check though whether these DOIs are correct, and if so, please add them to your BibTeX:
> 
> ```
> MISSING DOIs
> 
> - 10.2142/biophys.48.324 may be a valid DOI for title: Isca
> - 10.1017/cbo9780511617652.004 may be a valid DOI for title: Predictability: A problem partly solved
> ```

Hi @arfon ,

Thank you for your remark. It was already raised during the pre-review. Here was my answer to it: https://github.com/openjournals/joss-reviews/issues/2549#issuecomment-669902829 .

Basically the DOI Whedon proposes for Isca is wrong, but the description paper of Isca is correctly cited in our paper.
And the DOI for the Lorenz paper (*Predictability: A problem partly solved*) actually point to a re-edition but I want to cite the original one here. There is no DOI for the original one.

**Comment by jodemaey** [comment]
> And the DOI for the Lorenz paper (_Predictability: A problem partly solved_) actually point to a re-edition but I want to cite the original one here. There is no DOI for the original one.

Maybe I could cite the original article but then join to its citation the re-edition DOI so that people still get access to an electronic version.

**Comment by arfon** [comment]
> Maybe I could cite the original article but then join to it the citation the DOI of the re-edition so that people still get a link to an electronic version.

Yes, that sounds like a reasonable solution.

**Comment by jodemaey** [comment]
Hi @arfon ,

Sorry but after discussing with my co-authors, we don't think it is a good idea to put the DOI of one citation corresponding to another one.
Again we prefer to cite the original Lorenz 96 article since it is the one that gave its name to the model being referred to. It is also the seminal reference that is usually cited in our field (1163 times according to google).
It's a seminar proceeding that was never officially published and thus never got a DOI, however it is quite easy to find it on many place on the web with a quick google search.

If you agree with this, I will proceed and make the release.

Jonathan

**Comment by arfon** [comment]
> If you agree with this, I will proceed and make the release.

OK, this sounds reasonable. Please proceed.

**Comment by jodemaey** [comment]
Ok thanks, I will proceed as soon as possible.

**Comment by jodemaey** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02597/joss.02597/10.21105.joss.02597.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02597/joss.02597/10.21105.joss.02597.pdf) :page_facing_up: :point_left:

**Comment by jodemaey** [comment]
Hi @arfon ,

I've just merged the JOSS branch in master and released a new version [v0.2.1](https://github.com/Climdyn/qgs/releases/tag/v0.2.1) containing the paper and the modifications. Zenodi has automatically generated a new DOI: 10.5281/zenodo.4368844 that is already linked on the main page. Do I need to generate a specific archive for the article as you mentioned?

Also, please note that the online doc is currently not updated due to a [bug](https://github.com/executablebooks/jupyter-book/issues/1137#issuecomment-743993738) .

**Comment by arfon** [comment]
@whedon set 10.5281/zenodo.4368844 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.4368844" target="_blank">10.5281/zenodo.4368844</a> is the archive.

**Comment by arfon** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1016/j.physd.2015.07.006 is OK
- 10.5194/gmd-9-2793-2016 is OK
- 10.1007/s13351-018-8012-y is OK
- 10.1145/2833157.2833162 is OK
- 10.5194/npg-27-307-2020 is OK
- 10.1175/1520-0469(1963)020<0130:DNF>2.0.CO;2 is OK
- 10.1017/CBO9780511617652.004 is OK
- 10.1111/j.1600-0870.1984.tb00230.x is OK
- 10.1038/s41586-020-2649-2 is OK
- 10.1038/s41592-019-0686-2 is OK
- 10.3233/978-1-61499-649-1-87 is OK
- 10.5334/jors.151 is OK
- 10.1029/2019MS001652 is OK
- 10.1007/s10955-020-02525-z is OK
- 10.1002/qj.3594 is OK
- 10.1007/s00382-020-05313-3 is OK
- 10.5194/gmd-11-843-2018 is OK
- 10.3402/tellusa.v41i4.11842 is OK
- 10.1504/IJCSE.2009.029165 is OK

MISSING DOIs

- 10.2142/biophys.48.324 may be a valid DOI for title: Isca
- 10.1017/cbo9780511617652.004 may be a valid DOI for title: Predictability: A problem partly solved

INVALID DOIs

- 10.1175/1520-0469(1987)044%3C3282:TOMEAW%3E2.0.CO;2 is INVALID
- 10.1175/1520-0493(1982)110%3C1105:DOWRQS%3E2.0.CO;2 is INVALID
- 10.1002/(SICI)1097-024X(199606)26:6%3C635::AID-SPE26%3E3.0.CO;2-P is INVALID
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2004

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2004, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by arfon** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1342035657334648832) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/2005
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.02597](https://doi.org/10.21105/joss.02597)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by arfon** [comment]
@eviatarbach, @sadielbartholomew - many thanks for your reviews here. JOSS simply wouldn't work without the volunteer effort of folks like yourselves ✨ 

@jodemaey - your paper is now accepted into JOSS :zap::rocket::boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.02597/status.svg)](https://doi.org/10.21105/joss.02597)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.02597">
  <img src="https://joss.theoj.org/papers/10.21105/joss.02597/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.02597/status.svg
   :target: https://doi.org/10.21105/joss.02597
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.02597/status.svg)](https://doi.org/10.21105/joss.02597)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
